File: app/leetcode/No34.searchRange.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function (nums, target) {
    // 先用二分查找找出第一个元素的索引，然后用这个索引向左右两边进行扩散
    // let index = -1;
    // let l = 0, r = nums.length - 1;
    // while (l <= r) {
    //     const mid = l + parseInt((r - l) / 2);
    //     const num = nums[mid];
    //     if (num === target) {
    //         index = mid;
    //         break;
    //     }
    //     if (num < target) {
    //         l = mid + 1;
    //     } else {
    //         r = mid - 1;
    //     }
    // }
    // if (index === -1) {
    //     return [-1, -1];
    // }
    // let leftIndex = index;
    // let rightIndex = index;
    // while (leftIndex > 0 && nums[leftIndex - 1] === target) {
    //     leftIndex--;
    // }
    // while (rightIndex < nums.length - 1 && nums[rightIndex + 1] === target) {
    //     rightIndex++;
    // }
    // return [leftIndex, rightIndex];

    // 上面的算法在朴素的情况下复杂度确实是O(logN)的，但是对于数组中全是target的元素，复杂度退化为O(N)
    // 注意：下面的算法看起来像系统API，但是系统API可能出于通用性的考虑使用了线性查找
    function findFirstIndex(nums, target) {
        let l = 0, r = nums.length - 1;
        while (l <= r) {
            const mid = l + parseInt((r - l) / 2);
            const num = nums[mid];
            // 相等的时候并不跳出，而是继续向左边探测
            if (num === target) {
                r = mid - 1;
            } else if (num < target) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        if (l === nums.length || nums[l] !== target) {
            return -1;
        }
        return l;
    }
    function findLastIndex(nums, target) {
        let l = 0, r = nums.length - 1;
        while (l <= r) {
            const mid = l + parseInt((r - l) / 2);
            const num = nums[mid];
            if (num === target) {

            }
        }
    }
};

nums = [5, 7, 7, 8, 8, 10], target = 8
nums = [5, 7, 7, 8, 8, 10], target = 6
nums = [], target = 0
ret = searchRange(nums, target);
debugger

File: app/leetcode/No34.searchRange.py
from typing import List

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        # 类似 lfind 和 rfind
        # 暴力法，从前向后查找，再从后向前查找
        n = len(nums)
        res = [-1, -1]
        if n == 0:
            return res

        for i in range(n):
            if nums[i] == target:
                res[0] = i
                break
        for i in range(n-1, -1, -1):
            if nums[i] == target:
                res[1] = i
                break
        return res
    
    def searchRange2(self, nums: List[int], target: int) -> List[int]:
        # 解法 1 没有利用排序数组的特性
        res = [-1, -1]
        n = len(nums)
        if n == 0:
            return res
        
        # 先进行二分查找，找到目标值的索引，然后向左和向右进行线性扩散
        lo, hi = 0, n - 1
        index = -1
        while lo <= hi:
            mid = lo + (hi - lo) // 2
            v = nums[mid]
            if v == target:
                index = mid
                break
            if target > v:
                lo = mid + 1
            else:
                hi = mid - 1
        if index == -1:
            return res

        l = index
        while l >= 0 and nums[l] == v:
            res[0] = l
            l -= 1
        r = index
        while r < n and nums[r] == v:
            res[1] = r
            r += 1
        return res

    def searchRange3(self, nums: List[int], target: int) -> List[int]:                
        # 方法 2 改进了一些，但是在极端测试用例下会退化成 O(N)
        # 一个大数组所有元素都相同，需要找的元素刚好就在数组中
        # 可不可以一直使用二分呢？

        n = len(nums)
        res = [-1, -1]
        if n == 0:
            return res

        def findFirstIndex():
            lo, hi = 0, n - 1
            index = -1
            while lo <= hi:
                mid = lo + (hi - lo) // 2
                v = nums[mid]
                # 相等的时候继续尝试向左边探测
                if v == target:
                    index = mid
                    hi = mid - 1
                elif target > v:
                    lo = mid + 1
                else:
                    hi = mid - 1
            return index        

        def findLastIndex():
            lo, hi = 0, n - 1
            index = -1
            while lo <= hi:
                mid = lo + (hi - lo) // 2
                v = nums[mid]
                # 相等的时候继续尝试向右边探测
                if v == target:
                    index = mid
                    lo = mid + 1
                elif target > v:
                    lo = mid + 1
                else:
                    hi = mid - 1
            return index
        
        res[0] = findFirstIndex()
        res[1] = findLastIndex()
        return res


# [3,4]
# nums = [5,7,7,8,8,10]
# target = 8

# [-1,-1]
# nums = [5,7,7,8,8,10]
# target = 6

# [-1, -1]
# nums = []
# target = 0

nums = [2, 2]
target = 3

r = Solution().searchRange3(nums, target)
print(r)
